File: src/org/jgll/grammar/slotaction/NotMatchActions.java
package org.jgll.grammar.slotaction;

import java.util.List;

import org.jgll.grammar.Keyword;
import org.jgll.grammar.SlotAction;
import org.jgll.grammar.slot.BodyGrammarSlot;
import org.jgll.parser.GLLParserInternals;
import org.jgll.recognizer.GLLRecognizer;
import org.jgll.recognizer.RecognizerFactory;
import org.jgll.util.Input;
import org.jgll.util.hashing.CuckooHashSet;


public class NotMatchActions {

	  public static void fromGrammarSlot(BodyGrammarSlot slot, final BodyGrammarSlot ifNot) {

			slot.addPopAction(new SlotAction<Boolean>() {
				
				private static final long serialVersionUID = 1L;

				@Override
				public Boolean execute(GLLParserInternals parser, Input input) {
					GLLRecognizer recognizer = RecognizerFactory.contextFreeRecognizer();
					return recognizer.recognize(input, parser.getCurrentGSSNode().getInputIndex(), parser.getCurrentInputIndex(), ifNot);
				}
			});
		}
	  
		public static void fromKeywordList(BodyGrammarSlot slot, final List<Keyword> list) {
			
			if(list.size() == 1) {
				final Keyword s = list.get(0);
				
				slot.addPopAction(new SlotAction<Boolean>() {
					
					private static final long serialVersionUID = 1L;

					@Override
					public Boolean execute(GLLParserInternals parser, Input input) {
						return input.match(parser.getCurrentGSSNode().getInputIndex(), parser.getCurrentInputIndex(), s.getChars());
					}
				});
				
			} 
			
			else if(list.size() == 2) {
				final Keyword s1 = list.get(0);
				final Keyword s2 = list.get(1);
				
				slot.addPopAction(new SlotAction<Boolean>() {
					
					private static final long serialVersionUID = 1L;

					@Override
					public Boolean execute(GLLParserInternals parser, Input input) {
						int begin = parser.getCurrentGSSNode().getInputIndex();
						int end = parser.getCurrentInputIndex();
						return input.match(begin, end, s1.getChars()) ||
							   input.match(begin, end, s2.getChars())	;
					}
				});
			} 
			
			else {
				final CuckooHashSet<Keyword> set = new CuckooHashSet<>(Keyword.externalHasher);
				for(Keyword s : list) {
					set.add(s);
				}
				
				slot.addPopAction(new SlotAction<Boolean>() {

					private static final long serialVersionUID = 1L;
					
					@Override
					public Boolean execute(GLLParserInternals parser, Input input) {
						int begin = parser.getCurrentGSSNode().getInputIndex();
						int end = parser.getCurrentInputIndex() - 1;
						Keyword subInput = new Keyword("", input.subInput(begin, end));
						return set.contains(subInput);
					}
				});
			}
		}

	
}

File: src-scala/org/jgll_staged/grammar/slotaction/NotMatchActions.scala
package org.jgll_staged.grammar.slotaction

import java.util.List
import org.jgll_staged.grammar.Keyword
import org.jgll_staged.grammar.slot.BodyGrammarSlot
import org.jgll_staged.parser.GLLParserInternals
import org.jgll_staged.recognizer.GLLRecognizer
import org.jgll_staged.recognizer.RecognizerFactory
import org.jgll_staged.util.Input
import org.jgll_staged.util.hashing.CuckooHashSet
//remove if not needed
import scala.collection.JavaConversions._

object NotMatchActions {

  def fromGrammarSlot(slot: BodyGrammarSlot, ifNot: BodyGrammarSlot) {
    slot.addPopAction(new SlotAction[Boolean]() {

      private val serialVersionUID = 1L

      override def execute(parser: GLLParserInternals, input: Input): Boolean = {
        val recognizer = RecognizerFactory.contextFreeRecognizer()
        recognizer.recognize(input, parser.getCurrentGSSNode.getInputIndex, parser.getCurrentInputIndex, 
          ifNot)
      }
    })
  }

  def fromKeywordList(slot: BodyGrammarSlot, list: List[Keyword]) {
    if (list.size == 1) {
      val s = list.get(0)
      slot.addPopAction(new SlotAction[Boolean]() {

        private val serialVersionUID = 1L

        override def execute(parser: GLLParserInternals, input: Input): Boolean = {
          input.`match`(parser.getCurrentGSSNode.getInputIndex, parser.getCurrentInputIndex, s.chars)
        }
      })
    } else if (list.size == 2) {
      val s1 = list.get(0)
      val s2 = list.get(1)
      slot.addPopAction(new SlotAction[Boolean]() {

        private val serialVersionUID = 1L

        override def execute(parser: GLLParserInternals, input: Input): Boolean = {
          val begin = parser.getCurrentGSSNode.getInputIndex
          val end = parser.getCurrentInputIndex
          input.`match`(begin, end, s1.chars) || input.`match`(begin, end, s2.chars)
        }
      })
    } else {
      val set = new CuckooHashSet[Keyword](Keyword.externalHasher)
      for (s <- list) {
        set.add(s)
      }
      slot.addPopAction(new SlotAction[Boolean]() {

        private val serialVersionUID = 1L

        override def execute(parser: GLLParserInternals, input: Input): Boolean = {
          val begin = parser.getCurrentGSSNode.getInputIndex
          val end = parser.getCurrentInputIndex - 1
          val subInput = new Keyword("", input.subInput(begin, end))
          set.contains(subInput)
        }
      })
    }
  }
}
